prusti-interface/src/specs/checker.rs

```diff
@@ -126,12 +131,76 @@ impl<'v, 'tcx> Visitor<'tcx> for CheckPredicatesVisitor<'v, 'tcx> {
     }
 }
 
+/// Checks for the usage of models in non-specification code
+struct ModelUsageVisitor<'v, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    model_usages_in_non_spec_code: &'v mut Vec<Span>,
+}
+
+impl<'v, 'tcx> ModelUsageVisitor<'v, 'tcx> {
+    fn get_called_method(&self, expr: &'tcx hir::Expr<'tcx>) -> Option<hir::HirId> {
+        let maybe_method_def_id = self
+            .tcx
+            .typeck(expr.hir_id.owner)
+            .type_dependent_def_id(expr.hir_id);
+        if let Some(method_def_id) = maybe_method_def_id {
+            let maybe_local_def_id = method_def_id.as_local();
+            if let Some(local_def_id) = maybe_local_def_id {
+                let decl_hir_id = self.tcx.hir().local_def_id_to_hir_id(local_def_id);
+                return Some(decl_hir_id);
+            }
+        }
+        None
+    }
+
+    fn is_embedded_in_spec(&self, hir_id: hir::HirId) -> bool {
+        for (parent_local_def_id, _) in self.tcx.hir().parent_owner_iter(hir_id) {
```

Comment:
Why is the approach here different from `CheckPredicatesVisitor`? That is, why do we iterate over parents for calls to model methods rather than simply not recursing into spec functions in the first place?

Reply:
Good idea, I guess this will simplify the code

Reply:
I did a somewhat bigger refactoring of `SpecChecker`. The checker for type models now behaves like `CheckPredicatesVisitor`.